apps/src/lib/node/ledger/mod.rs

```diff
@@ -1,425 +1,214 @@
+            })),
+            Request::CheckTx(tx) => {
+                let r#type = match CheckTxType::from_i32(tx.r#type)
+                    .expect("received unexpected CheckTxType from ABCI")
+                {
+                    CheckTxType::New => MempoolTxType::NewTransaction,
+                    CheckTxType::Recheck => MempoolTxType::RecheckTransaction,
+                };
+                Ok(Response::CheckTx(self.mempool_validate(&*tx.tx, r#type)))
             }
-        }
+            Request::ListSnapshots(_) => {
+                Ok(Response::ListSnapshots(Default::default()))
+            }
+            Request::OfferSnapshot(_) => {
+                Ok(Response::OfferSnapshot(Default::default()))
+            }
+            Request::LoadSnapshotChunk(_) => {
+                Ok(Response::LoadSnapshotChunk(Default::default()))
+            }
+            Request::ApplySnapshotChunk(_) => {
+                Ok(Response::ApplySnapshotChunk(Default::default()))
+            }
+        };
+        tracing::info!(?rsp);
+        Box::pin(async move { rsp.map_err(|e| e.into()) }.boxed())
     }
+}
 
-    /// Simulate validation and application of a transaction.
-    fn dry_run_tx(&mut self, tx_bytes: &[u8]) -> Result<String> {
-        let mut gas_meter = BlockGasMeter::default();
-        let mut write_log = self.write_log.clone();
-        let result = protocol::apply_tx(
-            tx_bytes,
-            &mut gas_meter,
-            &mut write_log,
-            &self.storage,
+pub fn reset(config: config::Ledger) -> Result<(), shell::Error> {
+    shell::reset(config)
+}
+
+#[tokio::main]
+async fn run_shell(
+    config: config::Ledger,
+    abort_registration: AbortRegistration,
+) {
+    // Construct our ABCI application.
+    let service = Shell::new(&config.db, config::DEFAULT_CHAIN_ID.to_owned());
+
+    // Split it into components.
+    let (consensus, mempool, snapshot, info) = split::service(service, 1);
+
+    // Hand those components to the ABCI server, but customize request behavior
+    // for each category
+    let server = Server::builder()
+        .consensus(consensus)
+        .snapshot(snapshot)
+        .mempool(
+            ServiceBuilder::new()
+                .load_shed()
+                .buffer(10)
+                .service(mempool),
         )
-        .map_err(Error::TxError)?;
-        Ok(result.to_string())
-    }
+        .info(
+            ServiceBuilder::new()
+                .load_shed()
+                .buffer(100)
+                .rate_limit(50, std::time::Duration::from_secs(1))
+                .service(info),
+        )
+        .finish()
+        .unwrap();
 
-    /// Begin a new block.
-    fn begin_block(&mut self, hash: BlockHash, height: BlockHeight) {
-        self.gas_meter.reset();
-        self.storage.begin_block(hash, height).unwrap();
-    }
+    // Run the server with the shell
+    let future =
+        Abortable::new(server.listen(config.address), abort_registration);
```

Comment:
I'm not sure how to do this with the async inside the `Abortable`, but previously in `tendermint.rs:149`, we sent a special message to shut down Anoma node when Tendermint node goes down with `sender.send(AbciMsg::Terminate)`. I think it would be nice to do something similar

Reply:
I actually re-architected this whole function to ensure graceful shutdowns on panics in child theads.

Reply:
created #349